### ISSUE TITLE
Re-run only failed tests in reverted trunk commits

### ISSUE THREAD

**Issue by huydhn** [pytorch/test-infra#548]
### Pitch

This might be a controversial feature, but I think we could consider running only failed tests in reverted trunk commits.  The example I have here is as follows:

![Screen Shot 2022-08-16 at 16 46 20](https://user-images.githubusercontent.com/475357/185004480-786c94d3-2d55-44b0-82ff-6da62180949f.png)

1. #81789:  All green
1. #83422 and #83425:  These PRs broke ROCm tests in trunk
1. **#83425 is reverted later on**

Although we should re-run all the tests in trunk to make sure that nothing else break.  It's a reasonable tradeoff IMO to run only ROCm test in the last step.  My assumption is that the previous commit was all green and #83425 was green too with the only exception of ROCm test.

We revert commits quite often, so this should provide a good saving.  Also to be absolutely sure, we can skip over revert commits that run only a subset of tests when trying to promote them to viable/strict

### Solution

TBD

**Comment by ZainRizvi** [comment]
It's a nice idea.  The risk here would be if there was some kind of a race condition between the to-be-reverted PR and other PRs that the revert PR comes after, we wouldn't catch it.

E.g.:
1. PR 1 introduces a bug. A test starts failing
2. PR 2 gets force merged since clearly trunk is broken. It changes syntax in a way that the reverted PR cannot handle and break's some test
3. We don't discover the above break since the reverted PR doesn't actually run the test

Admittedly, the chances of the above scenario playing out are pretty low. But reverts tend to be really cheap compared to the cost of a normal PR (the entire workflow is only run once for a revert vs 3+ runs for a normal merge with land validation) so I'd lean towards the benefits not being worth the risk here.

**Comment by huydhn** [comment]
> It's a nice idea. The risk here would be if there was some kind of a race condition between the to-be-reverted PR and other PRs that the revert PR comes after, we wouldn't catch it.

Yeah, that's true. Thus, I'm thinking to myself that this is going to be controversial :P

It might work in practice because new commits are coming in quickly all the time.  And they are going to run the full CI anyway (and be promoted to viable/strict).  Any race condition coming from the revert will surface there right after the reverted commit.

**Comment by janeyx99** [comment]
We are also actively working towards minimizing reverts on trunk, so the tradeoffs of adding complexity + decreasing clarity in CI may not be worth doing.
